### ISSUE TITLE
Evaluation workflow

### ISSUE THREAD

**Issue by anastassiavybornova** [uscuni/simplification#200]
Add evaluation workflow (see `./evaluation/evaluate_h3cells.ipynb`) to compare `neatnet` output to other methods' output.

Current todos:

- [x] add computation of strokes (with momepy coins) to each network
- [x] add to stats: stroke count
- [x] add to stats: length of max stroke
- [x] add to stats: sum of intersecting stroke lengths
- [x] add to eval: weighting of cells by orig-manual multidim ranking
- [x] check in eval: which cells to downrank because of service road artifacts
- [x] add to eval: KDE plots for each evaluation variable (1x non-weighted, 1x-weighted)
- [x] norm colormaps to same lengths, ensure non-negativity

**Comment by anastassiavybornova** [comment]
xref #51

**Comment by anastassiavybornova** [comment]
@martinfleis  @jGaboardi 

some updates here on the `eval-updates` branch:

https://github.com/uscuni/simplification/tree/eval-updates/notebooks/evaluation 

i ran the notebook there `evaluation.ipynb` for auckland and for liege. will rerun once the updated manual data from @dancejod  is merged, and once we have decided on plot parameters etc. you can check all the plots in the folder for the following points TBD:

- [ ] does the current way of weighing cells make sense? (see below) @martinfleis  note that this is different than the conclusion from last time where we said to "just" normalize. (i realized that ofc if we divide all values by the same norm this is not "weighting" those values against each other)
- [ ] if so, do we also agree to completely discard the "non-weighted" version of plots, 
- [ ] any ideas on color schemes and alpha values etc. to make everything better visible? since we have 4 different methods to contrast against neatnet, it's hard to visualize all in 1 plot, so i made a subplots version also. 
- [ ] is it normal that reading in the data is taking sooo long? (almost 30mins for liege 😱 )

i'm cautiously optimistic that this looks pretty good, neatnet is not _always_ the best but _mostly at least as good, and often better_ vs. the other methods. and combined with a zoom-in evaluation of specific cases i think this will make a good case in point.

**re weighting**

Assuming that 

$$ var_{original} \geq var_{manual}  $$ 

holds for all cells, we compute the weight for variable in each cell as the ratio of its value in the original vs. the manually simplified data set,

$$ weight(cell) = \frac{var_{original}(cell)-var_{manual}(cell)}{var_{orig}(cell)}  $$

This will give $ weight(cell) \leq 1 $ for all cells.

**Comment by anastassiavybornova** [comment]
ps: there are some obvious reformatting/reshaping todos for some of the plots that currently don't look nice because of the sharex/sharey, but we can go into more detail once we have decided on weights and plotting parameters

**Comment by jGaboardi** [comment]
Looking through the plots. Magnificent.

**Comment by jGaboardi** [comment]
re: data loading time

```python
# get info on cells with to-be-revised data
deltas = gpd.read_file(f"../../revision/{fua}/deltas.gpkg")
```

That's the data load that's taking a crazy long time, right?

**Comment by anastassiavybornova** [comment]
@jGaboardi  no, reading in the deltas is not the issue - it's the `read_utils` function, particularly long for the two parenx subfolders... so weird

**Comment by martinfleis** [comment]
Can't reproduce that issue, it is immediate here. You can try purging the env and creating a fresh one by deleting `.pixi` and then doing `pixi install`.

**Comment by martinfleis** [comment]
I think that the new weighting is sensible. And agree to remove non-weighted plots. We need less, not more now.

Some of the figures are still hard to read, often due to osmnx using much higher range of values on y axis than the rest. We need to figure out how it should be best show. We will have a ton of these anyway (per city per variable) so maybe we can include the most representative sample in the paper and the rest in appendix?

**Comment by anastassiavybornova** [comment]
hello! @martinfleis  @jGaboardi just updated the evaluation (plot-producing) notebook on the `eval-updates` branch, as per your comments, and pushed updated (not final, not perfect) plots to a new `./plots/evaluation/{fua}` subfolder. 

also, after re-installing the pixi env, the data reading-in now takes only a few minutes instead of 30 😅 so thanks for the hint AND i'm still wondering what is going on

**Comment by jGaboardi** [comment]
> takes only a few minutes

Are we expecting it to still take minutes --> that's what you're wondering what's going on, right? Curious about your machine specs...

**Comment by martinfleis** [comment]
It should take ms no matter the machine.

**Comment by anastassiavybornova** [comment]
@martinfleis 
* to be collected here: ideas for statistical analysis of deltas
* the deltas to investigate - once #235 is merged: deltas to use for statistical analysis (and map plotting etc) will be in `./revision/{fua}/deltas_udpated.gpkg`

**Comment by martinfleis** [comment]
> ideas for statistical analysis of deltas

- two-sample Kolmogorov-Smirnov test
	- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ks_2samp.html#scipy.stats.ks_2samp
- Mann–Whitney _U_ test
	- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html#scipy.stats.mannwhitneyu
- Wilcoxon rank-sum statistic
	- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ranksums.html#scipy.stats.ranksums
- Cramér-von Mises test
	- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.cramervonmises_2samp.html#scipy.stats.cramervonmises_2samp
- Epps-Singleton
	- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.epps_singleton_2samp.html#scipy.stats.epps_singleton_2samp
- **Kullback–Leibler divergence**
	- aka relative entropy
	- a measure of how much a model [probability distribution](https://en.wikipedia.org/wiki/Probability_distribution "Probability distribution") Q is different from a true probability distribution P
	- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.entropy.html#scipy.stats.entropy
- **Jensen-Shannon distance**
	- https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.jensenshannon.html

**Comment by martinfleis** [comment]
> deltas to use for statistical analysis (and map plotting etc) will be in ./revision/{fua}/deltas_udpated.gpkg

@anastassiavybornova they are not. This only contains original vs manual. I don't think the deltas for other methods are saved, or am I wrong?

**Comment by jGaboardi** [comment]
@martinfleis I think you are wrong?

for example, see stuff in -- https://github.com/uscuni/simplification/tree/main/revision/1133

**Comment by martinfleis** [comment]
@jGaboardi but that geo package in there has a single layer which contains delta of original vs manual. Not of neatnet vs manual etc

**Comment by jGaboardi** [comment]
ah...

**Comment by martinfleis** [comment]
I was assuming those will be different layers in the gpkg but hold on, I have first need to figure out what I wanna do with that. In the meantime I can use it in-memory.

**Comment by anastassiavybornova** [comment]
@anastassiavybornova current evaluation subtasks:
- [ ] play around combining correlations from stats notebook
- [ ] play around with use case zoom-ins (provide bounding boxes)

**Comment by anastassiavybornova** [comment]
Hello hello! Some initial play-arounds with the stats viz: (see here on the eval-viz branch: https://github.com/uscuni/simplification/blob/eval-viz/notebooks/evaluation/statistical_comparison_viz.ipynb)

![Image](https://github.com/user-attachments/assets/d111bb34-ff92-4153-b00e-df54e171d1e5)

All 3 subplots contain exactly the same information, so the idea is to choose one of them and then optimize.

The different visual dimensions to consider would then be:
* city --> x or y axis in plot
* value of correlation --> y or x axis in plot
* method ---> different colors
* metric --> separate plots

will continue the playaround, but @martinfleis @jGaboardi  you're welcome to already drop your thoughts on the plots above and what could be done better!

**Comment by jGaboardi** [comment]
@anastassiavybornova This is great! I think the middle one (corr-y & city-x)  is the most easily interpretable (at least for me); absolutely clear with minimal mental effort.

**Comment by anastassiavybornova** [comment]
@jGaboardi @martinfleis  here is the entire correlation table info visualized in james' preferred style:

![Image](https://github.com/user-attachments/assets/7a4128d4-bc5e-4204-9d40-353a837912e1)

* i think we can drop at least one of the "stroke" vars, which will solve the blank last subplot 
* everything needs to be re-run with the updated wuhan data (once #250 is done, we should have a general manual-data cleanup i think)

**Comment by jGaboardi** [comment]
> i think we can drop at least one of the "stroke" vars, which will solve the blank last subplot

How about we have 2 figure of subplots?

- 2x2 for "graph measures"
- 1x3 for "continuity measure"

**Comment by anastassiavybornova** [comment]
> How about we have 2 figure of subplots?

yes either that, or we can use gridspec to make them all fit into 1 figure, either way is fine - but i think the more relevant question is to first decide whether we want to keep all 3 continuity measures at all

**Comment by jGaboardi** [comment]
> the more relevant question is to first decide whether we want to keep all 3 continuity measures at all

We may want to drop one of the three continuity plots because `stroke_length_max` and `stroke_length_sum` are so similar? Am I understanding correctly? Or something else?

**Comment by martinfleis** [comment]
Here is my suggestion. Don't want to push to avoid potential conflicts, but can do so if @anastassiavybornova permits.
![Image](https://github.com/user-attachments/assets/5cb57b0f-6f6d-418b-8cea-fe1cdd5e71c6)

```py
title_mapping = {
    "avg_degree": "Avg. degree",
    "coordinate_count": "Coordinate count",
    "edge_count": "Edge count",
    "edge_length": "Edge length",
    "stroke_count": "Stroke count",
    "stroke_length_max": "Stroke length (max)",
    "stroke_length_sum": "Stroke length (sum)",
}

fig, axs = plt.subplots(2, 4, sharex=True, sharey=True, figsize=(12, 6))

for i, metric in enumerate(eval_vars):
    ax = axs.flat[i]

    for method in methods_to_compare:
        ax.plot(
            combined.loc[metric][method],
            color=colors[method],
            linewidth=1,
            linestyle="dashed",
            marker="o",
            markersize=5,
            label=method,
        )
    ax.set_ylim(0, 1)
    if i == 6:
        ax.legend(loc="upper left", bbox_to_anchor=(1.15, 1.05), frameon=False)
    if i in range(4, 7):
        ax.set_xticks(list(xpos.values()))
        ax.set_xticklabels(list(xpos.keys()), rotation=90)
        # ax.set_xlabel("City")
    if i in [0, 4]:
        ax.set_ylabel(
            r"$\xi$ correlation"
        )
    ax.set_title(title_mapping[metric])
    ax.spines[['right', 'top']].set_visible(False)
axs.flat[-1].remove()
```

One more thing I'd do is to ensure that the order of cities in the legend matches the order of stroke length sum which is right next to it, so it is aligned.

This way we don't need to remove anything as the last square gets occupied by the legend.

**Comment by anastassiavybornova** [comment]
super cool @martinfleis thanks!! <3 i think it looks great now.

she allows pushing to the branch but she also already pushed it herself, it's in the updated eval viz notebook now.

are we happy enough with the plot for me to make a PR and then close this issue?

(will open a separate one for specific-example-finding, which is evaluation step 2; and will add the "reorder legend to fit max_stroke_length" to the list of things to do once we've run all the notebooks with revised_manual data)

**Comment by martinfleis** [comment]
go ahead and open the PR
